src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,133 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			int m = BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo;
+			if ( m > 0 )
+			{
+				char base[ MAX_STRING_CHARS ];
+				char rml[ MAX_STRING_CHARS ] = { 0 };
+				src = GetAttribute<Rocket::Core::String>( "src", "" );
+				Com_sprintf( base, sizeof( base ), "<img class='barbs' src='%s' />", src.CString() );
+
+				for ( int i=0; i<m; i++ )
+				{
+					Q_strcat( rml, sizeof( rml ), base );
+				}
+
+				// SetInnerRML deletes all child Elements => bring children up to date
+				children.clear();
+				SetInnerRML( rml );
+				for( int i=0; i<m; i++) {
+					children.push_back( GetChild( i ) );
+				}
+			}
+			else
+			{
+				children.clear();
+				SetInnerRML( "" );
+			}
+		}
+	}
+
+	void DoOnUpdate()
+	{
+		int barbs = cg.snap->ps.ammo;
+		int interval = GetInterval();
+
+		// start regenerating barb now
+		if ( barbs > numBarbs || ( barbs<numBarbs && numBarbs == BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo ) )
+		{
+			t0 = cg.time;
+			// sin(-pi/2) is minimal
+			offset = -M_PI_2;
+		}
+		// change regeneration speed
+		else if ( interval != regenerationInterval )
+		{
+			// black math magic to avoid sudden jumps in opacity
+			t0 = cg.time;
+			offset = asin( GetSin() ) - M_PI_2;
+			regenerationInterval = interval;
+		}
+		numBarbs = barbs;
+	}
+
+	void DoOnRender()
```

Comment:
This should go in DoOnUpdate() since the calls below aren't actual draw calls. They are just altering the state of the elements to be drawn.